feat: Enhance the RPM build

MichaelsJP opened the pull request:
### Pull Request Checklist
<!--- Please make sure you have completed the following items BEFORE submitting a pull request (put an x in each box
when you have checked you have done them): -->
- [ ] 1. I have [**rebased**][rebase] the latest version of the master branch into my feature branch and all conflicts
         have been resolved.
- [ ] 2. I have added information about the change/addition to functionality to the CHANGELOG.md file under the
         [Unreleased] heading.
- [ ] 3. I have documented my code using JDocs tags.
- [ ] 4. I have removed unnecessary commented out code, imports and System.out.println statements.
- [ ] 5. I have written JUnit tests for any new methods/classes and ensured that they pass.
- [ ] 6. I have created API tests for any new functionality exposed to the API.
- [ ] 7. If changes/additions are made to the ors-config.json file, I have added these to the [ors config documentation][config]
         along with a short description of what it is for, and documented this in the Pull Request (below).
- [ ] 8. I have built graphs with my code of the Heidelberg.osm.gz file and run the api-tests with all test passing
- [ ] 9. I have referenced the Issue Number in the Pull Request (if the changes were from an issue).
- [ ] 10. For new features or changes involving building of graphs, I have tested on a larger dataset
          (at least Germany), and the graphs build without problems (i.e. no out-of-memory errors).
- [ ] 11. For new features or changes involving the graphbuilding process (i.e. changing encoders, updating the
          importer etc.), I have generated longer distance routes for the affected profiles with different options
          (avoid features, max weight etc.) and compared these with the routes of the same parameters and start/end
          points generated from the current live ORS.
          If there are differences then the reasoning for these **MUST** be documented in the pull request.
- [ ] 12. I have written in the Pull Request information about the changes made including their intended usage
          and why the change was needed.
- [ ] 13. For changes touching the API documentation, I have tested that the API playground [renders correctly][api].

Fixes # .

### Information about the changes
- Key functionality added:
- Reason for change:

### Examples and reasons for differences between live ORS routes, and those generated from this pull request
-

### Required changes to ors config (if applicable)
-

[config]: https://GIScience.github.io/openrouteservice/installation/Configuration.html
[api]: https://gitlab.gistools.geog.uni-heidelberg.de/giscience/openrouteservice-infrastructure/ors-docs-api#test-new-ors-documentation
[rebase]: https://github.com/GIScience/openrouteservice/blob/master/CONTRIBUTE.md#pull-request-guidelines

jhaeu commented on .github/workflows/rpm-build.yml:
```diff
@@ -74,6 +74,41 @@ jobs:
           echo "RPM_PATH=${{ github.workspace }}/ors.rpm" >> "$GITHUB_ENV"
 
+      - name: Import the GPG key
+        uses: MichaelsJP/ghaction-import-gpg@v1.0
```
should we move this to GIScience?

MichaelsJP replied:
We still can in the future. But GIScience is already full of repos. I'd say we leave it at my space for now.

jhaeu commented on .github/workflows/rpm-build.yml:
```diff
@@ -74,6 +74,41 @@ jobs:
+          EOF
+          
+          echo "List loaded gpg keys in store"
+          gpg -K
```
use long options for better readability?

jhaeu commented on .rpm-packaging/test_rhel8_post_install_check.sh:
```diff
@@ -0,0 +1,58 @@
+)"
+
+# shellcheck disable=SC2034
+if [[ ! -d "$DIR" ]]; then DIR="$PWD"; fi
```
```suggestion
```

jhaeu commented on .rpm-packaging/test_rhel8_pre_install_check.sh:
```diff
@@ -0,0 +1,23 @@
+)"
+
+# shellcheck disable=SC2034
+if [[ ! -d "$DIR" ]]; then DIR="$PWD"; fi
```
```suggestion
```

jhaeu commented on .rpm-packaging/scripts/helper_functions.sh:
```diff
@@ -0,0 +1,262 @@
+
+    local result
+    local exit_code
+    result=$(docker exec -u root "$CONTAINER_NAME" bash -c "grep -Fx '$line' $path_to_file")
```
please test if correct: grep -Fx '$line'

MichaelsJP replied:
yes it its. the subshell is treating the single quotes differently. i added now escaped double quotes to make it clearer. because of all the encapsulated calls it's going through quite some shells

jhaeu commented on .rpm-packaging/test_rhel8_post_install_check.sh:
```diff
@@ -0,0 +1,58 @@
+find_owned_content "/opt/openrouteservice/*" "openrouteservice" "" 6 || exit 1
+find_owned_content "/opt/openrouteservice/*" "" "openrouteservice" 6 || exit 1
+find_owned_content "/opt/openrouteservice/*" "openrouteservice" "openrouteservice" 6 || exit 1
+find_owned_content "/opt/openrouteservice/*" "root" "root" 0 || exit 1
```
```suggestion
```

jhaeu commented on .rpm-packaging/test_rhel8_post_install_check.sh:
```diff
@@ -0,0 +1,58 @@
+find_owned_content "/opt/openrouteservice/*" "root" "root" 0 || exit 1
+find_owned_content "/opt/openrouteservice/*" "" "root" 0 || exit 1
+find_owned_content "/opt/openrouteservice/*" "root" "" 0 || exit 1
+find_owned_content "/opt/openrouteservice/*" "jboss" "jboss" 0 || exit 1
```
```suggestion
```

jhaeu commented on .rpm-packaging/ors-war.spec:
```diff
@@ -29,7 +31,7 @@ cp -f ors.war %{buildroot}%{ors_local_folder}/.war-files/%{ors_version}_ors.war
 
 %files
 # Allow 660 read and write for folders and 644 files for the ors group
```
```suggestion
# Allow 770 for read and write for files for the ors group
```

jhaeu commented on .rpm-packaging/ors-war.spec:
```diff
@@ -121,6 +123,8 @@ alternatives --set java $(readlink -f /etc/alternatives/jre_%{java_version})/bin
 
 # Set the correct permissions for the /opt/openrouteservice folder so that the ${ors_group} can read and write to it
 chown -R %{ors_user}:%{ors_group} %{ors_local_folder}
+# Set recursive 760 permissions for the /opt/openrouteservice folder so that the ${ors_group} can read and write to it
```
```suggestion
# Set recursive 770 permissions for the /opt/openrouteservice folder so that the ${ors_group} can read and write to it
```

jhaeu commented on .rpm-packaging/ors-war.spec:
```diff
@@ -121,6 +123,8 @@ alternatives --set java $(readlink -f /etc/alternatives/jre_%{java_version})/bin
+chmod -R 770 %{ors_local_folder}
 
 %postun
 # Uninstall routine if $1 is 0 but leave the opt folder
```
create test?

jhaeu reviewed (changes requested):
Excellent, much work! Just some small notes.

MichaelsJP committed:
refactor(docker): Rename the dockerfile

MichaelsJP committed:
feat(rpm): Add the correct rpm build procedure
A base script provides abstract check functions.
The three check scripts test the docker container before and after the installation and after removing to ensure that the installation environment is at all times as expected.
The spec file now enforces fixed file permissions.
The github workflow now signs, builds, installs, uninstalls and checks the rpm file at any given point.

MichaelsJP committed:
ci(github): Run the important workflows also in the version branch

jhaeu commented on .rpm-packaging/rhel8_post_uninstall_check.sh:
```diff
@@ -0,0 +1,50 @@
+# Check for owned content
+find_owned_content "/opt/openrouteservice/*" "openrouteservice" "" 0 || exit 1
+find_owned_content "/opt/openrouteservice/*" "" "openrouteservice" 0 || exit 1
+find_owned_content "/opt/openrouteservice/*" "openrouteservice" "openrouteservice" 0 || exit 1
```
```suggestion
```

MichaelsJP committed:
fix: Remove the unnecessary ownership check

Co-authored-by: jhaeu <138764906+jhaeu@users.noreply.github.com>

MichaelsJP committed:
feat(rpm): Allow the environment tests to run through even with failing checks

This allows the user to see the complete environment check even if one fails. That should allow for quicker bug fixing.

jhaeu reviewed (approved):
LGTM
